runner-review.sh

```diff
@@ -0,0 +1,333 @@
+
+## define global variables
+export DEBUG="set +x"                          ## disabled debug
+export execution_count=0                       ## total executions counter
+export DEFAULT_EXIT_CODE=255                   ## default exit-code
+export EXEC_NAME=$(basename $0)                ## script name 
+export OUTPUT_DIR="`pwd`"                      ## path for output & generated files 
+export TIME_SIGNUTURE="date +"%Y%m%d_%H%M%S""  ## current date&time
+
+
+## define local functions/methods 
+
+## print the summary report
+#  And exit the program with the most frequent return-code .
+#  in case no returned-codes (such case if the script killed while the first invocation) , then it will exit with default-return-code (predefined as 255)
+finalize() {
+    
+   frequent_rc=${DEFAULT_EXIT_CODE}
+   if [[ ! -z ${list_rc[@]} ]] ; then
+      
+      printf "|------Return Codes Summary-----|\n"
+      printf "|  Return-Code  |  Occurrence   |\n"
+      printf "|-------------------------------|\n"
+      echo ${list_rc[@]} | tr ' ' '\n' | sort | uniq -c | sort -rnk 1 | awk '{ printf "|\t%2d\t|\t%2d\t|\n", $2, $1 }' 
+      printf "|-------------------------------|\n"
+
+      frequent_rc=$(echo ${list_rc[@]} |tr ' ' '\n' | sort | uniq -c | sort -rnk 1 | head -1 | awk '{print $2}')
+   fi
+
+   exit ${frequent_rc}
+
+}
+
+## signal handler 
+#  catch system signals (SIGINT , SIGTERM) 
+# handler : call finalize
+# note : trap will not catch SIGKILL signal as OS doesn't allow it 
+trap finalize SIGINT SIGTERM
+
+
+
+## print script usage 
+usage() {
+
+	cat <<EOF
+Usage: ${EXEC_NAME} [OPTION].... [command [arg ...]]
+
+  -c COUNT            Number of times to run the given command
+  --failed-count N    Number of allowed failed command invocation attempts before giving up
+  --net-trace         For each failed execution, create a 'pcap' file with the network traffic during the execution.
+  --sys-trace         For each failed execution, create a log for each of the following values, measured during command execution
+					(disk io,+memory,processes/threads,cpu usage of the command network card package counters)
+  --call-trace        For each failed execution, add also a log with all the system calls ran by the command
+  --log-trace         For each failed execution, add also the command output logs (stdout, stderr)
+  --debug             Debug mode, show each instruction executed by the script
+EOF
+
+	exit ${DEFAULT_EXIT_CODE}
+}
+
+# generate network traffic capture by calling to tcpdump
+# default : collect 4000 packets for all the network interfaces formatted with default timestamp with extra verbose output 
+# note : those values is configurable and could be any value depend on the local machine/server and which details needed to capture from network traffic 
+exec_net_trace () {
+
+    export NET_PCAP="${OUTPUT_DIR}/net_trace_`${TIME_SIGNUTURE}`.pcap"
+    [[ ${NET_TRACE} ]] && /usr/sbin/tcpdump -i any -vvv -nnN -tttt -s0 -c 4000 -w ${NET_PCAP}
```

Comment:
`-i any` = listen to capture packets from all interfaces 
              we don't have any assumption about the network interfaces . 
`-vvv` = to print more verbose outputs like : ttl , identification, total length and options in an IP  packet  are  printed , more fields are printed from NFS reply packets ,etc...
            to collect more fields & data info . 
`-nnN` = protocols , port numbers , DNQ of host names will not be converted to names either .  
             to keep the sources/destinations/protocols without any names convert . 
`-tttt` = Print a timestamp in default format proceeded by date on each dump line . 
            to have more info about the packet timestmaps(date-time)
`-s0` = use the required length to catch whole packets(default is 68 bytes - is adequate for IP, ICMP, TCP and UDP but may truncate protocol information from name server and NFS packets )
`-c 4000` = capture only 4000 packets . 
                   In this case , it set to 4000packets , it could be ignore-able if the execution-command runs long time , so keep `tcpdump`  child-process run in background till the command execution finish , and then kill the `tcpdump` process . 
`-w ${NET_PCAP}` = capture and save the file in a .pcap format .